### ISSUE TITLE
DHCP Byte (order?) issue on SPARC CPUs

### ISSUE THREAD

**Issue by misullivan** [FreeRADIUS/freeradius-server#1954]
Its a byte ordering issue I think in the DHCP module - but may exist in the core as well.

Version seems to be 3.1.x (x=13 in my case) and not affecting 2.x

Received DHCP-Request of Id 2990dd58 from 10.10.0.20:68 to 10.10.0.3:67
	DHCP-Opcode = Client-Message
	DHCP-Hardware-Type = Ethernet
	DHCP-Hardware-Address-Length = 6
	DHCP-Hop-Count = 0
	DHCP-Transaction-Id = 697359704
	DHCP-Number-of-Seconds = 0
	DHCP-Flags = 0
	DHCP-Client-IP-Address = 0.0.0.0
	DHCP-Your-IP-Address = 0.0.0.0
	DHCP-Server-IP-Address = 0.0.0.0
	DHCP-Gateway-IP-Address = 0.0.0.0
	DHCP-Client-Hardware-Address = 00:11:22:33:44:55
	DHCP-Message-Type = DHCP-Request
	DHCP-Requested-IP-Address = 0.0.0.0
(0) Received code 1027 Id 697359704 from 10.10.0.20:68 to 10.10.0.3:67 length 250 
(0)   DHCP-Opcode = Client-Message
(0)   DHCP-Hardware-Type = Ethernet
(0)   DHCP-Hardware-Address-Length = 6
(0)   DHCP-Hop-Count = 0
(0)   DHCP-Transaction-Id = 697359704
(0)   DHCP-Number-of-Seconds = 0
(0)   DHCP-Flags = 0
(0)   DHCP-Client-IP-Address = 0.0.0.0
(0)   DHCP-Your-IP-Address = 0.0.0.0
(0)   DHCP-Server-IP-Address = 0.0.0.0
(0)   DHCP-Gateway-IP-Address = 0.0.0.0
(0)   DHCP-Client-Hardware-Address = 00:11:22:33:44:55
(0)   DHCP-Message-Type = DHCP-Request
(0)   DHCP-Requested-IP-Address = 0.0.0.0
(0) Trying sub-section dhcp <unknown> {...}
(0) Post-Auth-Type sub-section not found.  Ignoring.
ASSERT FAILED src/modules/proto_dhcp/dhcpd.c[384]: vp->vp_integer == 1
CAUGHT SIGNAL: Abort trap: 6

Putting in a temporary kludge at dhcpd.c line 304 of:

                if (vp->length == 1 && vp->vp_integer > 0xff) {
                        RDEBUG("ERROR: DHCP-Message-Type resolves to %d (0x%x) resetting to %d (0x%x)", vp->vp_integer, vp->vp_integer, vp->vp_integer >> 24, vp->vp_integer >> 24);
                        vp->vp_integer = vp->vp_integer >> 24;
                }

Gives the following output:

Received DHCP-Discover of Id f46460c4 from 10.10.0.252:67 to 10.10.0.3:67
	DHCP-Opcode = Client-Message
	DHCP-Hardware-Type = Ethernet
	DHCP-Hardware-Address-Length = 6
	DHCP-Hop-Count = 1
	DHCP-Transaction-Id = 4100219076
	DHCP-Number-of-Seconds = 25
	DHCP-Flags = 0
	DHCP-Client-IP-Address = 0.0.0.0
	DHCP-Your-IP-Address = 0.0.0.0
	DHCP-Server-IP-Address = 0.0.0.0
	DHCP-Gateway-IP-Address = 192.168.254.254
	DHCP-Client-Hardware-Address = 90:84:0d:f4:e0:53
	DHCP-Message-Type = DHCP-Discover
	DHCP-Parameter-Request-List = DHCP-Subnet-Mask
	DHCP-Parameter-Request-List = DHCP-Classless-Static-Route
	DHCP-Parameter-Request-List = DHCP-Router-Address
	DHCP-Parameter-Request-List = DHCP-Domain-Name-Server
	DHCP-Parameter-Request-List = DHCP-Domain-Name
	DHCP-Parameter-Request-List = DHCP-Domain-Search
	DHCP-Parameter-Request-List = DHCP-Site-specific-28
	DHCP-Parameter-Request-List = DHCP-LDAP
	DHCP-Parameter-Request-List = DHCP-NETBIOS-Name-Servers
	DHCP-Parameter-Request-List = DHCP-NETBIOS-Node-Type
	DHCP-DHCP-Maximum-Msg-Size = 1500
	DHCP-Client-Identifier = 0x0190840df4e053
	DHCP-IP-Address-Lease-Time = 7776000
	DHCP-Hostname = "oblivion"
(0) Received code 1025 Id -194748220 from 10.10.0.252:67 to 10.10.0.3:67 length 300 
(0)   DHCP-Opcode = Client-Message
(0)   DHCP-Hardware-Type = Ethernet
(0)   DHCP-Hardware-Address-Length = 6
(0)   DHCP-Hop-Count = 1
(0)   DHCP-Transaction-Id = 4100219076
(0)   DHCP-Number-of-Seconds = 25
(0)   DHCP-Flags = 0
(0)   DHCP-Client-IP-Address = 0.0.0.0
(0)   DHCP-Your-IP-Address = 0.0.0.0
(0)   DHCP-Server-IP-Address = 0.0.0.0
(0)   DHCP-Gateway-IP-Address = 192.168.254.254
(0)   DHCP-Client-Hardware-Address = 90:84:0d:f4:e0:53
(0)   DHCP-Message-Type = DHCP-Discover
(0)   DHCP-Parameter-Request-List = DHCP-Subnet-Mask
(0)   DHCP-Parameter-Request-List = DHCP-Classless-Static-Route
(0)   DHCP-Parameter-Request-List = DHCP-Router-Address
(0)   DHCP-Parameter-Request-List = DHCP-Domain-Name-Server
(0)   DHCP-Parameter-Request-List = DHCP-Domain-Name
(0)   DHCP-Parameter-Request-List = DHCP-Domain-Search
(0)   DHCP-Parameter-Request-List = DHCP-Site-specific-28
(0)   DHCP-Parameter-Request-List = DHCP-LDAP
(0)   DHCP-Parameter-Request-List = DHCP-NETBIOS-Name-Servers
(0)   DHCP-Parameter-Request-List = DHCP-NETBIOS-Node-Type
(0)   DHCP-DHCP-Maximum-Msg-Size = 1500
(0)   DHCP-Client-Identifier = 0x0190840df4e053
(0)   DHCP-IP-Address-Lease-Time = 7776000
(0)   DHCP-Hostname = "oblivion"
(0) ERROR: DHCP-Message-Type resolves to 16777216 (0x1000000) resetting to 1 (0x1)
(0) Trying sub-section dhcp DHCP-Discover {...}
(0)   dhcp DHCP-Discover {
(0)     update reply {
(0)       &DHCP-Message-Type = DHCP-Offer
(0)     } # update reply = noop
(0)     update reply {
(0)       &DHCP-Domain-Name-Server = 10.10.0.3
(0)       &DHCP-Domain-Name-Server = 192.168.250.2
(0)       &DHCP-Subnet-Mask = 255.255.255.0
(0)       &DHCP-Router-Address = 192.168.254.254
(0)       &DHCP-IP-Address-Lease-Time = 86400
(0)     } # update reply = noop
(0)     update control {
(0)       &Pool-Name := "wireless_254"
(0)     } # update control = noop
(0)     policy dhcp_sqlippool.post-auth {
(0)       update request {
(0)         EXPAND DHCP-%{DHCP-Client-Hardware-Address}
(0)            --> DHCP-90:84:0d:f4:e0:53
(0)         &User-Name = DHCP-90:84:0d:f4:e0:53
(0)         EXPAND %{DHCP-Client-Hardware-Address}
(0)            --> 90:84:0d:f4:e0:53
(0)         &Calling-Station-Id = 90:84:0d:f4:e0:53
(0)         EXPAND %{%{DHCP-Gateway-IP-Address}:-127.0.0.1}
(0)            --> 192.168.254.254
(0)         &NAS-IP-Address = 192.168.254.254
(0)         &Acct-Status-Type = Start
(0)       } # update request = noop
(0)       update control {
(0)         &Pool-Name := "wireless_254"
(0)       } # update control = noop
rlm_sql (sql): Reserved connection (0)
(0) dhcp_sqlippool: EXPAND %{User-Name}
(0) dhcp_sqlippool:    --> DHCP-90:84:0d:f4:e0:53
(0) dhcp_sqlippool: SQL-User-Name set to 'DHCP-90:84:0d:f4:e0:53'
(0) dhcp_sqlippool: EXPAND START TRANSACTION
(0) dhcp_sqlippool:    --> START TRANSACTION
(0) dhcp_sqlippool: Executing query: START TRANSACTION
(0) dhcp_sqlippool: EXPAND UPDATE radippool SET nasipaddress = '', pool_key = 0, callingstationid = '', username = '', expiry_time = NOW() WHERE expiry_time <= NOW() - INTERVAL 1 SECOND AND nasipaddress = '%{Nas-IP-Address}'
(0) dhcp_sqlippool:    --> UPDATE radippool SET nasipaddress = '', pool_key = 0, callingstationid = '', username = '', expiry_time = NOW() WHERE expiry_time <= NOW() - INTERVAL 1 SECOND AND nasipaddress = '192.168.254.254'
(0) dhcp_sqlippool: Executing query: UPDATE radippool SET nasipaddress = '', pool_key = 0, callingstationid = '', username = '', expiry_time = NOW() WHERE expiry_time <= NOW() - INTERVAL 1 SECOND AND nasipaddress = '192.168.254.254'
rlm_sql_mysql: Rows matched: 0  Changed: 0  Warnings: 0
(0) dhcp_sqlippool: EXPAND COMMIT
(0) dhcp_sqlippool:    --> COMMIT
(0) dhcp_sqlippool: Executing query: COMMIT
(0) dhcp_sqlippool: EXPAND START TRANSACTION
(0) dhcp_sqlippool:    --> START TRANSACTION
(0) dhcp_sqlippool: Executing query: START TRANSACTION
(0) dhcp_sqlippool: EXPAND SELECT framedipaddress FROM radippool WHERE pool_name = '%{control:Pool-Name}' AND (callingstationid = '%{Calling-Station-Id}' or callingstationid = '') ORDER BY (callingstationid <> '%{Calling-Station-Id}'), expiry_time LIMIT 1 FOR UPDATE
(0) dhcp_sqlippool:    --> SELECT framedipaddress FROM radippool WHERE pool_name = 'wireless_254' AND (callingstationid = '90:84:0d:f4:e0:53' or callingstationid = '') ORDER BY (callingstationid <> '90:84:0d:f4:e0:53'), expiry_time LIMIT 1 FOR UPDATE
(0) dhcp_sqlippool: Executing select query: SELECT framedipaddress FROM radippool WHERE pool_name = 'wireless_254' AND (callingstationid = '90:84:0d:f4:e0:53' or callingstationid = '') ORDER BY (callingstationid <> '90:84:0d:f4:e0:53'), expiry_time LIMIT 1 FOR UPDATE
(0) dhcp_sqlippool: Allocated IP 192.168.254.108
(0) dhcp_sqlippool: EXPAND UPDATE radippool SET nasipaddress = '%{NAS-IP-Address}', pool_key = '%{Calling-Station-Id}', callingstationid = '%{Calling-Station-Id}', username = '%{User-Name}', expiry_time = NOW() + INTERVAL 86400 SECOND WHERE framedipaddress = '192.168.254.108'
(0) dhcp_sqlippool:    --> UPDATE radippool SET nasipaddress = '192.168.254.254', pool_key = '90:84:0d:f4:e0:53', callingstationid = '90:84:0d:f4:e0:53', username = 'DHCP-90:84:0d:f4:e0:53', expiry_time = NOW() + INTERVAL 86400 SECOND WHERE framedipaddress = '192.168.254.108'
(0) dhcp_sqlippool: Executing query: UPDATE radippool SET nasipaddress = '192.168.254.254', pool_key = '90:84:0d:f4:e0:53', callingstationid = '90:84:0d:f4:e0:53', username = 'DHCP-90:84:0d:f4:e0:53', expiry_time = NOW() + INTERVAL 86400 SECOND WHERE framedipaddress = '192.168.254.108'
rlm_sql_mysql: Rows matched: 1  Changed: 1  Warnings: 0
(0) dhcp_sqlippool: EXPAND COMMIT
(0) dhcp_sqlippool:    --> COMMIT
(0) dhcp_sqlippool: Executing query: COMMIT
rlm_sql (sql): Released connection (0)
Need 5 more connections to reach 10 spares
rlm_sql (sql): Opening additional connection (5), 1 of 27 pending slots used
rlm_sql_mysql: Starting connect to MySQL server
rlm_sql_mysql: Connected to database 'radius' on Localhost via UNIX socket, server version 5.6.35, protocol version 10
(0)       [dhcp_sqlippool] = ok
(0)       if (ok) {
(0)       if (ok)  -> TRUE
(0)       if (ok)  {
(0)         update reply {
(0)           EXPAND %{reply:Framed-IP-Address}
(0)              --> 192.168.254.108
(0)           &DHCP-Your-IP-Address = 192.168.254.108
(0)         } # update reply = noop
(0)       } # if (ok)  = noop
(0)     } # policy dhcp_sqlippool.post-auth = ok
(0)     [updated] = updated
(0)   } # dhcp DHCP-Discover = updated
ASSERT FAILED src/modules/proto_dhcp/dhcpd.c[384]: vp->vp_integer == 1
CAUGHT SIGNAL: Abort trap: 6

Basically (and I'm having trouble finding the code that does the setting or I'd have patched it correctly..) it looks like when setting vp->vp_integer where vp->length = 1 the byte is going in as the first byte (MSB) of the integer because subsequent code doesn't know it's just a byte and not a 32bit int.

This is happening on a Netra T1-105 (Single UltraSPARC IIi 440MHz) running FreeBSD

**Comment by misullivan** [comment]
That works! (and just learned what a pieces of code I couldn't work out actually does...! Thanks!)

Now the only problem is to get the DHCP Server working.. it issues the offers, but the clients either ignore or reject the responses... but that is a conversation for somewhere other than here...

**Comment by arr2036** [comment]
http://freeradius.org/list/users.html works for DHCP too :)
